kf/bpf.go

```diff
@@ -534,56 +506,64 @@ func (b *BPF) UpdateArgs(ifaceName, direction string) error {
 }
 
 // Status of user program is running
-func (b *BPF) isRunning() (bool, error) {
-	// No user program or may be disabled
+func (b *BPF) isRunning() (bool, bool, error) {
+	userProgram := true
+	bpfProgram := false
+	var err error
+
+	// CmdStatus should check for user and BPF program
 	if len(b.Program.CmdStatus) > 1 {
 		cmd := filepath.Join(b.FilePath, b.Program.CmdStatus)
 
 		if err := assertExecutable(cmd); err != nil {
-			return false, fmt.Errorf("failed to execute %s with error: %v", b.Program.CmdStatus, err)
-		}
-
-		args := make([]string, 0, len(b.Program.StatusArgs)<<1)
+			userProgram = false
+		} else {
+			args := make([]string, 0, len(b.Program.StatusArgs)<<1)
 
-		for k, val := range b.Program.StatusArgs {
-			if v, ok := val.(string); !ok {
-				err := fmt.Errorf("status args is not a string for the ebpf program %s", b.Program.Name)
-				log.Error().Err(err).Msgf("failed to convert status args value into string for program %s", b.Program.Name)
-				return false, err
+			for k, val := range b.Program.StatusArgs {
+				if v, ok := val.(string); !ok {
+					err = fmt.Errorf("status args is not a string for the bpf program %s", b.Program.Name)
+					log.Warn().Err(err).Msgf("failed to convert status args value into string for program %s", b.Program.Name)
+				} else {
+					args = append(args, "--"+k+" ="+v)
+				}
+			}
+			prog := execCommand(cmd, args...)
+			var out bytes.Buffer
+			prog.Stdout = &out
+			prog.Stderr = &out
+			if err = prog.Run(); err != nil {
+				log.Warn().Err(err).Msgf("l3afd : Failed to execute %s", b.Program.CmdStatus)
+			}
+			outStr, errStr := out.String(), out.String()
+			if strings.EqualFold(outStr, bpfStatus) {
+				userProgram = true
+				bpfProgram = true
 			} else {
-				args = append(args, "--"+k+"="+v)
+				userProgram = false
+				log.Warn().Msgf("bpf program not running error - %s", errStr)
 			}
 		}
-
-		prog := execCommand(cmd, args...)
-		var out bytes.Buffer
-		prog.Stdout = &out
-		prog.Stderr = &out
-		if err := prog.Run(); err != nil {
-			log.Warn().Err(err).Msgf("l3afd/nf : Failed to execute %s", b.Program.CmdStatus)
-		}
-
-		outStr, errStr := out.String(), out.String()
-		if strings.EqualFold(outStr, bpfStatus) {
-			return true, nil
-		}
-
-		return false, fmt.Errorf("l3afd/nf : BPF Program not running %s", errStr)
+		return userProgram, bpfProgram, err
 	}
-
-	// No running user program and command status is not provided then return true
-	if !b.Program.UserProgramDaemon {
-		return true, nil
-	}
-
-	if err := b.VerifyProcessObject(); err != nil {
-		return false, errors.New("no process id found")
+	if len(b.Program.CmdStart) > 1 && b.Program.UserProgramDaemon {
+		if err := b.VerifyProcessObject(); err != nil {
+			userProgram = false
+			log.Warn().Err(err).Msgf("process object is not created for command start - %s", b.Program.CmdStart)
+		} else {
```

Comment:
I am not sure why this is an if / else clause.  Can't we have one way to verify the process is running when it is running as a daemon? Why can't we do all the needed checks for this condition in the IsProcessRunning function?

Reply:
Updated IsProcessRunning is called inside VerifyProcessObject.